docs: add end2end example on creating a basic text-classification dataset

plaguss opened the pull request:
<!-- Thanks for your contribution! As part of our Community Growers initiative 🌱, we're donating Justdiggit bunds in your name to reforest sub-Saharan Africa. To claim your Community Growers certificate, please contact David Berenstein in our Slack community or fill in this form https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR includes 2 features towards the #4178 issue.
- A tutorial for the creation of a `FeedbackDataset` for `text-classification`.
- A new script has been added to run the notebooks automatically, via end2end.yml workflow.

Closes #4179 and #4220

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [ ] `sphinx-autobuild` (read [Developer Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation) for more details)

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

plaguss committed:
docs: first tutorial for end2end examples on feedback dataset for text classification

plaguss committed:
Merge branch 'develop' of github.com:argilla-io/argilla into docs/end2end-text-classification

plaguss committed:
docs: update with task templates and push to argilla hf hub

plaguss committed:
feat: parametrize notebook to run via papermill

plaguss committed:
feat: script to run end2end notebooks

plaguss committed:
refactor: grab apikey based on quickstart image

plaguss committed:
feat: initial version of end2end examples from ci/cd

davidberenstein1957 commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,124 @@
+    raise ValueError("No token found. Please set HF_HUB_ACCESS_TOKEN environment variable.")
+
+
+def main(api_url: str = "http://localhost:6900", api_key: Optional[str] = None, hf_token: Optional[str] = None) -> None:
```
I prefer to also use the default `ARGILLA_API_URL`when it is None?

davidberenstein1957 commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,124 @@
+            print(f"Removed output notebook: {self.dst_filename.stem}")
+
+
+def get_argilla_api_key() -> Optional[str]:
```
Don`t we normally use `ARGILLA_API_KEY` here?

plaguss replied:
I wasn't sure of the env var names, and assumed the ones [here](https://github.com/argilla-io/argilla/blob/develop/docker/scripts/start_quickstart_argilla.sh) should be chosen, but I'll update those

davidberenstein1957 commented on .github/workflows/end2end-examples.yml:
```diff
@@ -0,0 +1,76 @@
+        run: |
+          echo "ARGILLA_SEARCH_ENGINE=opensearch" >> "$GITHUB_ENV"
+          echo "Configure opensearch engine"
+      - name: Run end2end examples 📈
```
Because of running this every time, do you think we can filter ut a but and only run it when there are changes to `src` or `examples.py`? Also, perhaps we can use a subset of the `datasets` and/or setup a  persistent cache for the `datasets` and set this equal to our "Cache pip 👜" step?

davidberenstein1957 replied:
This might be relevant in other places we download 'datasets'  for our cache.

plaguss replied:
Will update that

davidberenstein1957 commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,124 @@
+        "hf_token": hf_token,
+    }
+
+    examples = [
```
perhaps it is better to do this with `glob` and select everything in our folder as examples?

plaguss replied:
I wasn't sure whether we could add other files here, but I think that's better

plaguss replied:
@davidberenstein1957 I think we would like to have these run in a specific order. For that should we name them using something that can be sorted? (maybe just a number at the end works fine)

plaguss committed:
feat: apply comments from code review

Commit:
feat: update workflow to run depending on the result from check-repo-files

Commit:
feat: update naming convention for examples to allow sorting them before running

Commit:
refactor: update glob call and sort notebooks before running

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

Commit:
chore: remove deleted file

Commit:
Merge branch 'docs/end2end-text-classification' of https://github.com/argilla-io/argilla into docs/end2end-text-classification

Commit:
Merge branch 'develop' of https://github.com/argilla-io/argilla into docs/end2end-text-classification

Commit:
feat: add telemetry for the tutorial

Commit:
refactor: set python3 as the default kernel for running the notebooks

davidberenstein1957 committed:
tests: updated en2end workflow

davidberenstein1957 committed:
chore: added file changes for end2end tests
chore: changed notebook to trigger workflow

davidberenstein1957 committed:
chore: updated server launch command

davidberenstein1957 committed:
chore: updated credentials

davidberenstein1957 committed:
chore: add HF_HUB_ACCESS_TOKEN

davidberenstein1957 committed:
chore: add .png for workflows

davidberenstein1957 committed:
docs: updated end2end tutorials

davidberenstein1957 committed:
chore: updated starting reference

davidberenstein1957 committed:
chore: updated relative path to end2end examples

davidberenstein1957 committed:
chore: updated check-repo-files to hit updates in correct example folder

davidberenstein1957 committed:
docs: updated phrasing
docs: added general workflow image

davidberenstein1957 committed:
docs: updated phrasing

davidberenstein1957 committed:
docs: renamed files for sequential runnign

davidberenstein1957 committed:
docs: added updated data model

davidberenstein1957 committed:
docs: update user management section

davidberenstein1957 committed:
docs: added extra section on users and workspaces

kursathalat committed:
docs: Adding `metadata` to a `text-classification` dataset (#4313)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

Please include a summary of the changes and the related issue. Please
also include relevant motivation and context. List any dependencies that
are required for this change.
- To do: add links to previous tutorial in Pull Dataset section and to
subsequent tutorial in Conclusion section.

Closes #4181 

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [ ] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

davidberenstein1957 committed:
docs: slight updates phrasing

frascuchon commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,119 @@
+    raise ValueError("No token found. Please set HF_HUB_ACCESS_TOKEN environment variable.")
+
+
+def main(api_url: Optional[str] = None, api_key: Optional[str] = None, hf_token: Optional[str] = None) -> None:
```
`api_url` and `api_key` are not used. I can imagine they must be used to setup the Argilla client for each tutorial, right?

plaguss replied:
They were initially, but [this](https://github.com/argilla-io/argilla/pull/4208/files#diff-bbe8ef3e2a3d64e39e8d9dae60c31a354b1f47980ecab20a4113f73ac1bb2cd1R174-R179) addition from @davidberenstein1957 seems good enough. What do you think?

frascuchon replied:
If the environment override is not so difficult I think is better and more reusable to keep notebooks setup agnostic and let the runner configure the setup. 

So, I prefer to pass everything the notebook needs for the setup as parameters in the papermill call and let it set up the environment.

plaguss replied:
👍

frascuchon commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,119 @@
+
+EXAMPLES_FOLDER = (
+    Path(__file__).parent.parent / "docs/_source/tutorials_and_integrations/tutorials/feedback/end2end_examples"
+)
```
I would say to define this as script argument:

```bash
end2end_examples "docs/_source/tutorials_and_integrations/tutorials/feedback/end2end_examples"
```

plaguss replied:
I'll tackle this

plaguss replied:
It has the following arguments now:

```bash
❯ python scripts/end2end_examples.py --help

 Usage: end2end_examples.py [OPTIONS]

 Run the end2end example notebooks. If no arguments are passed, it will try to get the api_key and the hf_token from the
 environment variables.

╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --api-url                TEXT  [default: http://localhost:6900]                                                       │
│ --api-key                TEXT  [default: argilla.apikey]                                                              │
│ --hf-token               TEXT  [default: None]                                                                        │
│ --examples-folder        TEXT  [default: docs/_source/tutorials_and_integrations/tutorials/feedback/end2end_examples] │
│ --help                         Show this message and exit.                                                            │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

The expected path is set as default but can be overriden

frascuchon commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,119 @@
+            papermill.execute_notebook(
+                str(self.src_filename), str(self.dst_filename), parameters=self.parameters, kernel_name="python3"
+            )
+            print(f"✅  {self.src_filename.stem}")
```
As a future improvement, to be sure the notebook does what it needs to do, we could have a step that executes a check_notebook_*.py file where all the assert/checks are done. But we can have a deep conversation on that without blocking this PR

plaguss replied:
Sure, added issue to keep track of this: #4324

frascuchon commented on .github/workflows/package.yml:
```diff
@@ -74,6 +74,28 @@ jobs:
+    name: Run end2end tests
+    uses: ./.github/workflows/end2end-examples.yml
+    needs: check_repo_files
+    if: needs.check_repo_files.outputs.end2endChanges == 'true'
```
We can start with this condition, but these tests won't be executed all the time. We can schedule a nightly execution on dev, and one execution when tag/release is going to happen.

plaguss replied:
👍

Commit:
Merge branch 'develop' into docs/end2end-text-classification

Commit:
feat: update parameter cells and some comments in notebooks

Commit:
feat: update parameters from code review

frascuchon commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,130 @@
+    # Name of the output notebook that will be removed after running the examples
+    output_notebook = "output_notebook.ipynb"
+
+    examples_folder = ROOT_FOLDER / examples_folder
```
can we just assume that the examples folder if relative, is to the script working directory? I think we can just:
```bash
cd argilla-repo
python scripts/end2end_examples.py
```
Since the working dir is the repo root, the script call should work. 

Having this computed value depending on where the script is placed normally brings errors.

plaguss replied:
@frascuchon, https://github.com/argilla-io/argilla/pull/4208/commits/8b91a3cf4e2748346d60cc4c5df3812f995b5b51 should adress this. I let in the docstrings for the script how to run it just in case.

Commit:
fix: remove root folder reference from the script, warn that the script must be run from the root

kursathalat committed:
docs: Adding `vectors` to a text-classification dataset (#4338)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

Please include a summary of the changes and the related issue. Please
also include relevant motivation and context. List any dependencies that
are required for this change.

Closes #4182 

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [ ] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: sdiazlor <sdiazlor@gmail.com>

frascuchon commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,128 @@
+                parameters=self.parameters,
+                kernel_name="python3",
+            )
+            print(f"✅  {self.src_filename}")
```
Does it make sense to align these prints with `rich`? I think we're already using it in other places. 

cc @davidberenstein1957

davidberenstein1957 replied:
I agree

plaguss replied:
Do we have some place where we call the print from rich specifically? I've seen [this](https://github.com/argilla-io/argilla/blob/develop/src/argilla/cli/rich.py) for the CLI, but there's no specific print from `Console`. I just wanted some simple print to keep track of the process because it's for internal use, but let me know and I'll update it accordingly

frascuchon commented on scripts/end2end_examples.py:
```diff
@@ -0,0 +1,128 @@
+    # Name of the output notebook that will be removed after running the examples
+    output_notebook = "output_notebook.ipynb"
+
+    output_folder = examples_folder.parent / "end2end_outputs"
```
I would say to:

1. Output results in the working dir as default
2. allow to setup this folder as a command argument

plaguss replied:
I updated that to write directly to tmp. If we need the resulting notebooks from papermill we can just change that, for the moment we don't use them and we can simply remove them at the end

kursathalat committed:
docs: Update records with `Responses` and `Suggestions` (#4326)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

To do: 
- Responses will be handled.
- Links to the next tutorials will be added.

Closes #4183 

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [ ] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

Commit:
refactor: write output file to tmpdir

sdiazlor committed:
docs: example on assigning record to your team for a text classification dataset (#4342)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

Please include a summary of the changes and the related issue. Please
also include relevant motivation and context. List any dependencies that
are required for this change.

Closes #4184 

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [x ] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

davidberenstein1957 committed:
docs: updated references

davidberenstein1957 committed:
docs: added reference to parpermill description

davidberenstein1957 committed:
docs: updated tabs

davidberenstein1957 committed:
docs: marked cells as parameters

davidberenstein1957 committed:
docs: assign records

davidberenstein1957 committed:
chore: add utils module

davidberenstein1957 committed:
docs: assign records

davidberenstein1957 committed:
Merge branch 'develop' into docs/end2end-text-classification

davidberenstein1957 committed:
chore: updated cell id

davidberenstein1957 committed:
chore: added kernelspec info

davidberenstein1957 committed:
chore: updated add vectors

davidberenstein1957 committed:
docs: updated suggestions and responses

kursathalat committed:
docs: Tutorial on fine-tuning for a text-classification dataset (#4348)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

Please include a summary of the changes and the related issue. Please
also include relevant motivation and context. List any dependencies that
are required for this change.

Closes #4187 

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [ ] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

davidberenstein1957 committed:
docs: updated token

davidberenstein1957 committed:
docs: added updated to end2end training

davidberenstein1957 committed:
chore: review notebook validator

sdiazlor committed:
docs: example on filtering and querying records for a text classification dataset (#4350)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

Please include a summary of the changes and the related issue. Please
also include relevant motivation and context. List any dependencies that
are required for this change.

Closes #4185 

**Type of change**

(Remember to title the PR according to the type of change)

- [ x] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [ x] `sphinx-autobuild` (read [Developer
Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation)
for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

davidberenstein1957 committed:
docs: updated end2end to running examples

davidberenstein1957 committed:
docs: updated tutorials and images

davidberenstein1957 committed:
docs: updated message used

davidberenstein1957 committed:
docs: added reference to dataset
docs: changed to spacy training instead of trf

davidberenstein1957 committed:
docs: push with local model instead

davidberenstein1957 committed:
docs: updated practical guides image
